### PR TITLE
refine multi cards logic

### PR DESCRIPTION
**Pull request by shippingwang**
A little words before
PaddlePaddle will start num_cards * num_workers subprocess to deal with data
and will start num_cards subprocess(distinguish by **rank** ) to train the model.
like: when set two cards and two num_workers, it will start 1(main process)+2*2 +2 = 7 process on the operating system.

This PR
1. add a decorate to refine some functions which should only call once in the main process, like: save/load_ckpt
2. release subprocess data when exiting the main process to avoid manual kill them.
3. change the interface to get current process rank and the number of all ranks, will auto judge parallel mode, no need to pass an arg: parallel
4. support multi-cards validate, In eval mode, will reduce sum all cards process metrics, usually include top1 and top5 acc, so that users can get the final acc in the rank0 log without manual calculate multi-cards acc.

### CODE REVIEW

File: paddlevideo/modeling/heads/base.py [hunk 1]
```diff
@@ -80,10 +80,17 @@ def loss(self, scores, labels, **kwargs):
         #XXX: F.crossentropy include logsoftmax and nllloss 
         loss = self.loss_func(scores, labels, **kwargs)
         avg_loss = paddle.mean(loss)
-        softmax_out = F.softmax(scores)         
+        softmax_out = F.softmax(scores)
```
Comment (LielinJiang) [inline paddlevideo/modeling/heads/base.py:83]: 似乎没有必要

File: paddlevideo/utils/dist_utils.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import paddle
+import paddle.distributed as dist
+
+def get_dist_info():
+    world_size = dist.get_world_size()
+    rank = dist.get_rank()
+    return rank, world_size
+
+def main_only(func):
```
Comment (LielinJiang) [inline paddlevideo/utils/dist_utils.py:24]: 函数名称有点怪怪的，不过也想不到太好的

**Review by LielinJiang — APPROVED** [review]

File: paddlevideo/modeling/heads/base.py (hunk 1 above)
Comment (shippingwang) [reply paddlevideo/modeling/heads/base.py:83]: fc之后不过softmax，算的top1/5也是一样的呗

File: paddlevideo/utils/dist_utils.py (hunk 2 above)
Comment (shippingwang) [reply paddlevideo/utils/dist_utils.py:24]: 对，要不叫master_only...

File: main.py [hunk 3]
```diff
@@ -53,7 +50,9 @@ def main():
     args = parse_args()
 
     cfg = get_config(args.config, overrides=args.override)
-    if args.parallel:
+    _, world_size = get_dist_info()
+    parallel = world_size != 1
+    if parallel:
         paddle.distributed.init_parallel_env()
 
     model = build_model(cfg.MODEL)
```
Comment (littletomatodonkey) [inline main.py:58]: parallel的话，需要model = paddle.DataParallel(model)， 这边没看到，如果没写的话可以补充下

File: paddlevideo/modeling/framework/recognizers/base.py [hunk 4]
```diff
@@ -60,13 +60,13 @@ def forward_valid(self, imgs):
         pass
 
 
-    def forward(self, imgs, labels=None, return_loss=True, **kwargs):
+    def forward(self, imgs, labels=None, return_loss=True, reducesum=False, **kwargs):
```
Comment (littletomatodonkey) [inline paddlevideo/modeling/framework/recognizers/base.py:63]: reduce_sum更好理解一些？

File: paddlevideo/tasks/train.py [hunk 5]
```diff
@@ -41,7 +41,7 @@ def train_model(model,
     dataset = dataset if isinstance(dataset, (list, tuple)) else [dataset]
     #build data loader, refer to the field ```DATASET``` in the configuration for more details.
     batch_size = cfg.DATASET.get('batch_size', 2)
-    places = paddle.CUDAPlace(paddle.distributed.ParallelEnv().dev_id)
+    places = paddle.set_device('gpu')
```
Comment (littletomatodonkey) [inline paddlevideo/tasks/train.py:44]: 这块感觉可以加一下cpu逻辑，也比较方便
```
paddle.set_device('gpu' if use_gpu else 'cpu')
```

File: paddlevideo/tasks/train.py [hunk 6]
```diff
@@ -121,12 +120,15 @@ def train_model(model,
...
+            test()
+
         if metric_list['top1'].avg > best:
             best = metric_list['top1'].avg
             opt_state_dict = optimizer.state_dict()
             opt_name = cfg['OPTIMIZER']['name']
-            paddle.save(opt_state_dict, f"{opt_name}.pdopt")
-            paddle.save(model.state_dict(), "best.pdparams")
+            save(opt_state_dict, f"{opt_name}.pdopt")
+            save(model.state_dict(), "best.pdparams")
```
Comment (littletomatodonkey) [inline paddlevideo/tasks/train.py:131]: 这块保存要不统一一下名字？直接写best感觉有点粗暴

File: paddlevideo/tasks/train.py [hunk 7]
```diff
@@ -90,15 +90,14 @@ def train_model(model,
...
-        
         # learning scheduler step
         lr.step()
 
         ips = "ips: {:.5f} instance/sec.".format(batch_size * metric_list["batch_time"].count / metric_list["batch_time"].sum)
         log_epoch(metric_list, epoch, "train", ips)
 
 
-        if validate:
+        def test():
```
Comment (littletomatodonkey) [inline paddlevideo/tasks/train.py:100]: 这快要不重命名为validate()或者evaluate()吧

File: main.py (hunk 3 above)
Comment (shippingwang) [reply main.py:58]: 有的，这里

File: paddlevideo/modeling/framework/recognizers/base.py (hunk 4 above)
Comment (shippingwang) [reply paddlevideo/modeling/framework/recognizers/base.py:63]: ok

File: paddlevideo/tasks/train.py (hunk 5 above)
Comment (shippingwang) [reply paddlevideo/tasks/train.py:44]: cpu不验证了，但是可以写上：）

File: paddlevideo/tasks/train.py (hunk 6 above)
Comment (shippingwang) [reply paddlevideo/tasks/train.py:131]: 🉑️ 这块需要优化下

File: paddlevideo/tasks/train.py (hunk 7 above)
Comment (shippingwang) [reply paddlevideo/tasks/train.py:100]: 🉑️
